lms/djangoapps/certificates/api.py

```diff
@@ -268,6 +268,38 @@ def generate_certificate_task(user, course_key, generation_mode=None):
     return _generate_certificate_task(user, course_key, generation_mode)
 
 
+def allow_certificate_available_date(course_overview, certificate_status=None):
+    """
+    Check that certificate_available_date is enabled to display in settings and status.
+
+    Args:
+        course_overview (CourseOverview object): ID associated with the course
+        certificate_status (dict object or None): certificate status if it's downloadable/auditing/generating
+
+    Returns:
+        True if the certificate enabled in settings, display behavior and certificate status is valid,
+        and certificate_available_date is not None else False
+    """
+    certificate_status = (
+        CertificateStatuses.is_passing_status(certificate_status['status'])
+        if certificate_status
+        else True
+    )
+    if settings.FEATURES.get("ENABLE_V2_CERT_DISPLAY_SETTINGS"):
+        display_behavior_is_valid = (
+            course_overview.certificates_display_behavior == CertificatesDisplayBehaviors.END_WITH_DATE
+        )
+    else:
+        display_behavior_is_valid = True
+
+    return (
+        not certificates_viewable_for_course(course_overview) and
+        certificate_status and
+        display_behavior_is_valid and
+        course_overview.certificate_available_date
+    )
```

Comment:
This might be a long comment but I was digging around to see how generic we can make this and use this commonly with reference to my previous comment about a reusable function to check the display behavior setting here and in other places.

I was looking at different APIs that return `certificate_available_date` 
So, while looking around I saw many usages of checking the display behavior in detail. These APIs are mentioned in https://github.com/mitodl/mitxonline/issues/1257#issuecomment-1346463078 so I will try not to re-write all that.

In my analysis, i noticed that these APIs at different points call https://github.com/openedx/edx-platform/blob/96b42271e3041f9ce4cd4a53566b7907569e902c/lms/djangoapps/courseware/views/views.py#L1070 in combination with different checks on certificate status. This function internally calls other functions that deeply involve student and certificate status which i really think we don't need.

e.g. in the selected lines here if you check  `not certificates_viewable_for_course(course_overview)` in detail it also checks different types of certificate state checks that we might not need.

Here are my testing results:
1. This worked partially I think, e.g. if disabling the waffle switch `certificates.auto_certificate_generation` which turns off the display behavior setting in the studio, I still get the `certificate_available_date` in the course details API although the behavior is disabled in the studio.
2. The above point made me more curious about digging into the APIs in detail that returns `cert_data` in response and I think if we were to create this utility function and use it everywhere we'll have to make a lot of changes in different pieces of code that will increase the scope of this and we might end up breaking something that we don't intend to.

**My suggestion:**

I think for this specific API, we should only check for `ENABLE_V2_CERT_DISPLAY_SETTINGS` and `certificates.auto_certificate_generation`. These are the exact setting that enables certificate display behavior in studio and these should also enable the addition of `certificate_available_date` in the course details API.

What i think is that we should remove this function and do something like:

```
if settings.FEATURES.get("ENABLE_V2_CERT_DISPLAY_SETTINGS") and auto_certificate_generation_enabled():
        display_behavior_is_valid = (
            course_overview.certificates_display_behavior == CertificatesDisplayBehaviors.END_WITH_DATE
        )
    else:
        display_behavior_is_valid = True

// If the above is True, add the `certificate_available_date` in the course details API. 
```

and undo the current refactoring we did below in `certificate_downloadable_status`. This would make the PR simpler, touching the least parts of the other functionality.

Reply:
I think your suggestion makes sense to me.

Reply:
I’ve already removed this. Please review it again